src/devices/bus/sg1000_exp/sp400.cpp

```diff
@@ -0,0 +1,249 @@
+// license: BSD-3-Clause
+// copyright-holders: Charles MacDonald, Devin Hill, Fabio Dalla Libera
+
+/*
+    license:BSD-3-Clause
+    Charles MacDonald
+    Devin Hill
+    Fabio Dalla Libera
+
+    Sega SP-400 Plotter
+
+    6805 dumped using https://github.com/charlesmacd/HD6805_Reader
+
+    Plotter uses ALPS Micrographic Printer DPG 1302 mechanism.
+
+    IC list:
+
+    * M5L8035LP (Mitsubishi clone of I8035)
+
+    P1 is the deserialized output data
+    P2.4 is the input serial data
+    P2.5 is busy (output to the computer)
+    P2.6 is the output strobe
+    P2.7 is busy (input from the 6805)
+
+    * HD6805V (Hitachi chip, pin compatible with MC6805P2, with 4K of ROM)
+
+    port A connected to the panel buttons
+        "By analyzing the disassembly I found out that three pins (PA5,6,7 of the 6805) are used to set
+        different configurations. The resolution can be changed, probably to support different mechanics.
+        The baud rate is 4800."
+    port B connected to the motors
+    port C connected to the pen and misc signals
+    port D connected to M5L8035LP parallel data
+*/
+
+// this is currently incomplete: missing paper output and pen sensor, everything else is working
+
+#include "emu.h"
+#include "sp400.h"
+
+#include "cpu/mcs48/mcs48.h"
+#include "cpu/m6805/m68705.h"
+
+namespace {
+
+class sp400_printer_device : public device_t,
+	public device_sk1100_printer_port_interface
+{
+public:
+	sp400_printer_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
+
+	static constexpr feature_type unemulated_features() { return feature::PRINTER; }
+
+	virtual void device_add_mconfig(machine_config &config) override ATTR_COLD;
+	virtual void device_start() override ATTR_COLD;
+	virtual void device_reset() override ATTR_COLD;
+	virtual void device_reset_after_children() override ATTR_COLD;
+
+	const tiny_rom_entry * device_rom_region() const override ATTR_COLD;
+	virtual ioport_constructor device_input_ports() const override ATTR_COLD;
+
+	virtual void input_data (int state) override { m_data = state; }
+	virtual int output_busy () override { return m_busy; }
+
+private:
+	void dser_map(address_map &map);
+	void dser_p1_w(uint8_t data);
+	uint8_t dser_p2_r();
+	void dser_p2_w(uint8_t data);
+
+	uint8_t mot_pa_r();
+	void mot_pb_w(uint8_t data);
+	uint8_t mot_pc_r();
+	void mot_pc_w(uint8_t data);
+	uint8_t mot_pd_r();
+	void update_pen_state(u8 data);
+	u8 pen_sensor() { return 0; }  // not currently hooked up
+
+	required_device<i8035_device> m_dsercpu;     // "deserializer cpu"
+	required_device<m6805_hmos_device> m_motcpu; // "motor cpu"
+
+	int m_data, m_busy;
+	int m_dserdataout, m_dserstrobe;
+	int m_motbusy, m_motPenUp, m_motPenDown;
+	int m_pendown;
```

Comment:
Please use `snake_case` consistently rather than this mix of `snakecase` and `llamaCase`.

Also, please use explicitly sized integers for things that need to be registered for save states.